controllers/ocirepository_controller.go

```diff
@@ -0,0 +1,909 @@
+		resErr   error
+		metadata = sourcev1.Artifact{}
+	)
+
+	// Run the sub-reconcilers and build the result of reconciliation.
+	for _, rec := range reconcilers {
+		recResult, err := rec(ctx, obj, &metadata, tmpDir)
+		// Exit immediately on ResultRequeue.
+		if recResult == sreconcile.ResultRequeue {
+			return sreconcile.ResultRequeue, nil
+		}
+		// If an error is received, prioritize the returned results because an
+		// error also means immediate requeue.
+		if err != nil {
+			resErr = err
+			res = recResult
+			break
+		}
+		// Prioritize requeue request in the result.
+		res = sreconcile.LowestRequeuingResult(res, recResult)
+	}
+
+	r.notify(ctx, oldObj, obj, res, resErr)
+
+	return res, resErr
+}
+
+// reconcileSource fetches the upstream OCI artifact metadata and content.
+// If this fails, it records v1beta2.FetchFailedCondition=True on the object and returns early.
+func (r *OCIRepositoryReconciler) reconcileSource(ctx context.Context, obj *sourcev1.OCIRepository, metadata *sourcev1.Artifact, dir string) (sreconcile.Result, error) {
+	ctxTimeout, cancel := context.WithTimeout(ctx, obj.Spec.Timeout.Duration)
+	defer cancel()
+
+	options := r.craneOptions(ctxTimeout)
+
+	// Generate the registry credential keychain either from static credentials or using cloud OIDC
+	keychain, err := r.keychain(ctx, obj)
+	if err != nil {
+		e := serror.NewGeneric(
+			fmt.Errorf("failed to get credential: %w", err),
+			sourcev1.AuthenticationFailedReason,
+		)
+		conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+		return sreconcile.ResultEmpty, e
+	}
+	options = append(options, crane.WithAuthFromKeychain(keychain))
+
+	if obj.Spec.Provider != sourcev1.GenericOCIProvider {
+		auth, authErr := r.oidcAuth(ctxTimeout, obj)
+		if authErr != nil && !errors.Is(authErr, oci.ErrUnconfiguredProvider) {
+			e := serror.NewGeneric(
+				fmt.Errorf("failed to get credential from %s: %w", obj.Spec.Provider, authErr),
+				sourcev1.AuthenticationFailedReason,
+			)
+			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+			return sreconcile.ResultEmpty, e
+		}
+		if auth != nil {
+			options = append(options, crane.WithAuth(auth))
+		}
+	}
+
+	// Generate the transport for remote operations
+	transport, err := r.transport(ctx, obj)
+	if err != nil {
+		e := serror.NewGeneric(
+			fmt.Errorf("failed to generate transport for '%s': %w", obj.Spec.URL, err),
+			sourcev1.AuthenticationFailedReason,
+		)
+		conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+		return sreconcile.ResultEmpty, e
+	}
+	if transport != nil {
+		options = append(options, crane.WithTransport(transport))
+	}
+
+	// Determine which artifact revision to pull
+	url, err := r.getArtifactURL(obj, options)
+	if err != nil {
+		if _, ok := err.(invalidOCIURLError); ok {
+			e := serror.NewStalling(
+				fmt.Errorf("failed to determine the artifact address for '%s': %w", obj.Spec.URL, err),
```

Comment:
This results in:
> failed to determine the artifact address for 'oci://ghcr.io/stefanprodan/manifests/podinfo:aa': URL must not contain a tag; remove ':aa'

I think it'd be more appropriate to to say something like `URL validation failed: URL must not contain ...` or something better, since this and the generic error below are handling two different types of failures.